serdect/src/slice.rs

```diff
@@ -32,91 +36,39 @@ where
     S: Serializer,
     T: AsRef<[u8]>,
 {
-    if serializer.is_human_readable() {
-        crate::serialize_hex::<_, _, true>(value, serializer)
-    } else {
-        value.as_ref().serialize(serializer)
+    common::serialize_hex_upper_or_bin(value, serializer)
+}
+
+struct UpperBound;
+
+impl LengthCheck for UpperBound {
+    fn length_check(buffer_length: usize, data_length: usize) -> bool {
+        buffer_length >= data_length
+    }
+    fn expecting(
+        formatter: &mut fmt::Formatter<'_>,
+        data_type: &str,
+        data_length: usize,
+    ) -> fmt::Result {
+        write!(
+            formatter,
+            "{} with a maximum length of {}",
+            data_type, data_length
+        )
     }
 }
 
 /// Deserialize from hex when using human-readable formats or binary if the
 /// format is binary. Fails if the `buffer` is smaller then the resulting
 /// slice.
-pub fn deserialize_hex_or_bin<'de, D>(buffer: &mut [u8], deserializer: D) -> Result<&[u8], D::Error>
+pub fn deserialize_hex_or_bin<'de, D>(buffer: &mut [u8], deserializer: D) -> Result<(), D::Error>
```

Comment:
@fjarri running into problems with this change when upgrading this crate.

This API in particular is for deserializing variable-length data, however removing the `&[u8]` return value means the amount of data that was actually deserialized was lost.

`buffer` is intended to hold a backing buffer which is as large as the deserialized data can possibly be, but it may be smaller, and the return value is there to return the actual amount of data deserialized.

Reply:
Made a tracking issue: #1322